src/main/java/internet/store/dao/jdbc/ShopCartDaoJdbcImpl.java

```diff
@@ -0,0 +1,141 @@
+package internet.store.dao.jdbc;
+
+import internet.store.dao.ShopCartDao;
+import internet.store.exception.DataProcessingException;
+import internet.store.lib.Dao;
+import internet.store.model.Product;
+import internet.store.model.ShoppingCart;
+import internet.store.util.ConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ShopCartDaoJdbcImpl implements ShopCartDao {
+    @Override
+    public Optional<ShoppingCart> getByUserId(Long userId) {
+        String query = "SELECT user_id, cart_id, id, product_name, price FROM users "
+                + "JOIN shopping_cart using(user_id) "
+                + "LEFT JOIN shopping_cart_product USING(cart_id) "
+                + "LEFT JOIN products ON products.id = shopping_cart_product.product_id "
+                + "WHERE user_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, userId);
+            ResultSet resultSet = statement.executeQuery();
+            ShoppingCart shoppingCart = new ShoppingCart(userId);
+            List<Product> productList = new ArrayList<>();
+            if (resultSet.next()) {
+                shoppingCart.setCartId(resultSet.getLong("cart_id"));
+            }
+            do {
+                Long productId = resultSet.getObject("id", Long.class);
+                if (productId != null) {
+                    productList.add(retrieveProductDataFromDB(resultSet));
+                }
+            } while (resultSet.next());
+            shoppingCart.setProducts(productList);
+            return Optional.of(shoppingCart);
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Shopping cart for user "
+                    + userId + " not found", ex);
+        }
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart shoppingCart) {
```

Comment:
In this method if shopping cart already has some products we ignore it. I'm not sure about the logic, should we create shopping cart with some products in it, but in my implementation I add this products to new shopping cart

Reply:
My application has the saem result as you put but probably different steps. I have three steps: 
(1)new ShopCart is predefined with changes (= with addition of new product), 
2) all products in mySQL(only!)  from shop cart are deleted 
(3) all products (old and new one) are created based on ShopCar from p.1.
So I get the same result as you assumed in your application.